gateways/core/calendar/calendar.go

```diff
@@ -0,0 +1,212 @@
+	// the combination of these rules and dates combine to form a set of date times.
+	// NOTE: functionality currently only supports EXDATEs, but in the future could be expanded.
+	Recurrence []string `json:"recurrence,omitempty" protobuf:"bytes,3,rep,name=recurrence"`
+}
+
+type calendar struct {
+	gatewayConfig *gateways.GatewayConfig
+	registeredCalendarSignals []calSchedule
+}
+
+func (c *calendar) RunGateway(cm *apiv1.ConfigMap) error {
+	CheckAlreadyRegistered:
+		for scheduleKey, scheduleValue := range cm.Data {
+			var cal *calSchedule
+			err := yaml.Unmarshal([]byte(scheduleValue), &cal)
+			if err != nil {
+				c.gatewayConfig.Log.Error().Str("artifact", scheduleKey).Err(err).Msg("failed to parse calendar schedule")
+				return err
+			}
+			c.gatewayConfig.Log.Info().Interface("artifact", *cal).Msg("calendar schedule")
+			for _, registeredCalendarSignal := range c.registeredCalendarSignals {
+				if cmp.Equal(registeredCalendarSignal, cal) {
+					c.gatewayConfig.Log.Warn().Str("schedule", cal.Schedule).Str("interval", string(cal.Interval)).Str("recurrences", strings.Join(cal.Recurrence, ",")).Msg("calendar schedule is already registered")
+					goto CheckAlreadyRegistered
+				}
+			}
+			c.registeredCalendarSignals = append(c.registeredCalendarSignals, *cal)
+			go c.startCalenderSchedule(cal)
+		}
+	return nil
+}
+
+func (c *calendar) startCalenderSchedule(cal *calSchedule) error {
+	schedule, err := c.resolveSchedule(cal)
+	if err != nil {
+		return fmt.Errorf("failed to resolve calendar schedule. Err: %+v", err)
+	}
+	exDates, err := common.ParseExclusionDates(cal.Recurrence)
+	if err != nil {
+		return fmt.Errorf("failed to resolve calendar schedule. Err: %+v", err)
+	}
+
+	var next Next
+	next = func(last time.Time) time.Time {
+		nextT := schedule.Next(last)
+		nextYear := nextT.Year()
+		nextMonth := nextT.Month()
+		nextDay := nextT.Day()
+		for _, exDate := range exDates {
+			// if exDate == nextEvent, then we need to skip this and get the next
+			if exDate.Year() == nextYear && exDate.Month() == nextMonth && exDate.Day() == nextDay {
+				return next(nextT)
+			}
+		}
+		return nextT
+	}
+
+	lastT := time.Now()
+	for {
+		t := next(lastT)
+		timer := time.After(time.Until(t))
+		log.Printf("expected next calendar event %s", t)
+		select {
+		case tx := <-timer:
+			lastT = tx
+			c.gatewayConfig.Log.Info().Msg("event sending")
+			event := metav1.Time{Time: t}
+			payload, err := event.Marshal()
+			if err != nil {
+				return fmt.Errorf("failed to marshal event. Err: %+v", err)
+			} else {
+				// dispatch the event to gateway transformer
+				_, err = http.Post(fmt.Sprintf("http://localhost:%s", c.gatewayConfig.TransformerPort), "application/octet-stream", bytes.NewReader(payload))
+				if err != nil {
+					return fmt.Errorf("failed to dispatch the event. Err: %+v", err)
+				}
+				c.gatewayConfig.Log.Info().Msg("event dispatched to gateway transformer")
```

Comment:
what about creating `kubernetes.Events` for this? this would enable us to create a log of all the events dispatched...

Reply:
Do you have something in mind like a gateway dispatches an event by creating kubernetes.Event with "type" of particular configuration of that gateway and sensors then listens to such event/s? 

This does make the ecosystem completely kubernetes native and remove need for a pub-sub system that sit between gateway and sensor. Is etcd storage a right choice to store events for production grade deployment of argo-events?

Another approach we can take is to make framework opinionated about internal pub-sub and use NATS streaming. Gateways will publish to NATS streaming cluster and sensors will subscribe to it. This makes gateways and sensors completely decoupled from each other. Some of the advantages NATS streaming has to offer are "Historical message replay by subject" , "At-least-once-delivery" and "Message/event persistence". We can also provide out-of-box support for indexing and monitoring using elastic search and prometheus if an user wishes to use them. 

Thoughts?

Reply:
my thought here wasn't about "production" grade messaging layer. it was more for a temporary reporting/accounting store. I don't think using Kubernetes Events here is a good idea, however i do believe an internal pub/sub interface/mechanism (with persistent and guarantees) would be. I do however think possible recording meta-events about the state of gateways/sensors through Kubernetes events would be interesting. for example, publishing an event when a gateway reaches a running state.. this seems to be the kind of behavior other default Kubernetes objects do during their lifecycle (e.g. pods -> pulling images, creating container, etc.)